src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -303,3 +303,57 @@
             [(conj errors {:index row-index, :error (ex-message e)})
              created-rows]))))
      rows)))
+
+(defn- pk-values->filters [table-id pk-values]
+  ;; check that all pk-values have the same shape
+  (when-not (apply = (map (comp set keys) pk-values))
+    (throw (ex-info (trs "Inconsistent row selection.")
+                    {:status-code 400, :pk-values pk-values})))
+  (let [given-pks (set (keys (first pk-values)))
+        table-pks (into {} (map (juxt :name :id) (filter #(= (:semantic_type %) :type/PK) (:fields (first (table/with-fields [(Table table-id)]))))))]
```

Comment:
Not that this is super important **here** since this code probably won't be invoked often, but this isn't a super efficient way to get the `:name` and `:id` of the PK Field(s) for a given Table in question, either in terms of performance or code. So let me show you how to do it the nice way

The current code:

1. fetches an entire Table object in one query
2. fetches*all* of its Fields (the entire object)
3. Filters out any Fields that aren't PKs in Clojure-land
4. Takes just the `:name` and `:id` columns for each Field and building a map

Now whenever you're doing filtering or key selection **in Clojure-land** you should stop and ask yourself whether this is something that can be done in the App DB instead. With Toucan it's easy to do something like

```clj
(db/select [Field :id :name] :table_id table-id)
```

which is the equivalent of 

```sql
SELECT id, name FROM field WHERE table_id = ?
```

For filtering on `semantic_type` we have `metabase.db.util/isa` that generates an appropriate HoneySQL clause to do the equivalent of `isa?` in DB-land.

Furthermore Toucan has `db/select-field->id` which you can use to select a specific field and return a map of values of that field to the corresponding values of `id` for that row, since this is something that you actually want to do pretty often.

Put that together and you can write this as:

```clj
(db/select-field->id :name Field
  {:where [:and
           [:= :table_id table-id]
           (mdb.u/isa :semantic_type :type/PK)]})
```

This does the exact same thing but does the equivalent of 

```sql
SELECT name, id 
FROM FIELD
WHERE table_id = ?
  AND semantic_type IN ('type/PK')
```

which is a lot more efficient.